collections.js

```diff
@@ -118,8 +118,49 @@ LinkedList.prototype.find = function(value) {
     node = node.next;
   }
   return null;
+};
+
+//------------------------------------------------------------------------------
+// Queue, a simple queue implementation.
+// The head of the queue is at the beginning of the backing array.
+
+function Queue() {
+  this.head_ = 0;
+  this.headLimit_ = 1000;
+  this.sliceMin_ = 10;
+  this.q_ = [];
 }
 
+Queue.prototype = {
+  enqueue: function(item) {
+    this.q_.push(item);
+    return this;
+  },
+
+  dequeue: function() {
+    var result = this.q_[this.head_];
+    delete this.q_[this.head_];
+    this.head_ = this.head_ + 1;
+    if (this.head_ >= this.headLimit_ ||
+        this.head_ > this.sliceMin_ && this.head_ > (this.q_.length - this.head_)) {
```

Comment:
Sure, on big queues (past say 2000 in length).

Reworked the code a bit to maintain a public length property that makes this condition a bit more readable.